include/wil/registry.h

```diff
@@ -28,6 +28,166 @@ namespace wil
     size_t const max_registry_value_name_length = 16383;
 
+#ifdef WIL_ENABLE_EXCEPTIONS
+    inline wil::unique_hkey open_registry_key(HKEY hkey, PCWSTR subkey, DWORD options = 0, REGSAM samDesired = KEY_READ)
+    {
+        wil::unique_hkey openKey;
+        THROW_IF_FAILED(HRESULT_FROM_WIN32(::RegOpenKeyExW(hkey, subkey, options, samDesired, &openKey)));
+        return openKey;
+    }
+
+    // TODO: can we make any of these accesses simpler?
+    inline wil::unique_hkey create_registry_key(
+        HKEY hkey,
+        PCWSTR subkey,
+        PWSTR classType = nullptr,
+        DWORD options = REG_OPTION_NON_VOLATILE,
+        REGSAM samDesired = KEY_READ,
+        SECURITY_ATTRIBUTES* securityAttributes = nullptr,
+        _Out_ DWORD* disposition = nullptr)
+    {
+        wil::unique_hkey openKey;
+        THROW_IF_FAILED(HRESULT_FROM_WIN32(::RegCreateKeyExW(hkey, subkey, 0, classType, options, samDesired, securityAttributes, &openKey, disposition)));
+        return openKey;
+    }
+#endif // WIL_ENABLE_EXCEPTIONS
+
+    inline HRESULT get_registry_dword_nothrow(HKEY hkey, LPCWSTR subkey, LPCWSTR regValueName, _Out_ DWORD* value) noexcept
+    {
+        DWORD size{sizeof(value)};
+        // TODO: support internal registry functions?
+        // TODO: support additional flags, like RRF_ZEROONFAILURE?
+        // TODO: convert to HRESULT?
+        const auto hr = HRESULT_FROM_WIN32(::RegGetValueW(hkey, subkey, regValueName, RRF_RT_DWORD, nullptr, value, &size));
+        return hr;
+    }
+
+    /// @cond
+    namespace details
+    {
+        template <typename err_policy, typename result = err_policy::result>
+        inline result set_registry_dword(HKEY hkey, LPCWSTR subkey, LPCWSTR regValueName, DWORD value)
+        {
+            // TODO: support little endian and big endian?
+            const auto hr = HRESULT_FROM_WIN32(::RegSetKeyValueW(hkey, subkey, regValueName, REG_DWORD, &value, sizeof(value)));
+            return err_policy::HResult(hr);
+        }
+    }
+    /// @endcond
+
+    constexpr auto* set_registry_dword_nothrow = details::set_registry_dword<err_returncode_policy>;
+    constexpr auto* set_registry_dword_failfast = details::set_registry_dword<err_failfast_policy>;
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+
+    constexpr auto* set_registry_dword = details::set_registry_dword<err_exception_policy>;
+
+    /// @cond
+    namespace details
+    {
+#if _HAS_CXX17 && defined(_OPTIONAL_)
+        template <typename Result>
+        struct optional_registry_policy
+        {
+            typedef std::optional<Result> result;
+            __forceinline static result NotFound() { return std::nullopt; }
+        };
+#endif // _HAS_CXX17 && defined(_OPTIONAL_)
+
+        template<typename Result>
+        struct required_registry_policy
+        {
+            typedef Result result;
+            __forceinline static result NotFound()
+            {
+                THROW_HR(HRESULT_FROM_WIN32(ERROR_NOT_FOUND));

```

Comment:
WIL already has macros taking Win32 error code as input.
`THROW_WIN32(ERROR_NOT_FOUND)` would be simpler here.